### PR TITLE
Keep active on restart

### PR DESCRIPTION
**Pull request by smk762**
When resetting assets, this will preserve active coins.
Closes https://github.com/KomodoPlatform/atomicDEX-Desktop/issues/1544 and https://github.com/KomodoPlatform/atomicDEX-Desktop/pull/1632

### CODE REVIEW

File: src/core/atomicdex/pages/qt.settings.page.cpp [hunk 1]
```diff
@@ -606,12 +585,84 @@ namespace atomic_dex
             }
         };
 
-        functor_remove(std::move(wallet_cfg_path));
+        if (fs::exists(wallet_cfg_path))
+        {
+            nlohmann::json wallet_config_json_data;
+            nlohmann::json active_list = nlohmann::json::array();
```
Comment (Milerius) [inline src/core/atomicdex/pages/qt.settings.page.cpp:591]: I think this container is suboptimal for what you are trying to achieve.

Take a look at <https://en.cppreference.com/w/cpp/container/unordered_set> - an `unordered_set<std::string>` seems more appropriate and the following usage <https://en.cppreference.com/w/cpp/container/unordered_set/contains>.

File: src/core/atomicdex/pages/qt.settings.page.cpp [hunk 2]
```diff
@@ -606,12 +585,84 @@ namespace atomic_dex
...
+            fs.close();
+
+            //! set active coins again
+            for (auto&& [key, value]: wallet_config_json_data.items())
+            {
+                if (std::find(active_list.begin(), active_list.end(), key) != active_list.end())
+                {
+                    value["active"] = true;
+                }
+            }
```
Comment (Milerius) [inline src/core/atomicdex/pages/qt.settings.page.cpp:629]: Here by using the `unordered_set<string>` approach you will reduce drastically your complexity <https://www.bigocheatsheet.com/>

you have now an O(n)² complexity.

you will be at O(n) complexity afterwards.

**Review by Milerius — CHANGES_REQUESTED** [review]

File: src/core/atomicdex/pages/qt.settings.page.cpp [hunk 3]
```diff
@@ -606,12 +585,83 @@ namespace atomic_dex
...
+            fs.open(QIODevice::ReadOnly | QIODevice::Text);
+
+            //! Read Contents
+            wallet_config_json_data = nlohmann::json::parse(QString(fs.readAll()).toStdString());
+            fs.close();
+
+            //! set active coins again
+            for (auto&& key: active_list)
+            {
+                if (active_list.contains(key))
```
Comment (Milerius) [inline src/core/atomicdex/pages/qt.settings.page.cpp:625]: If you iterate through the active list, the key is already present so the if condition is useless here.

Please also rename `active_list` to `active_coins_registry`

**Review by Milerius — CHANGES_REQUESTED** [review]

File: src/core/atomicdex/pages/qt.settings.page.cpp [hunk 4]
```diff
@@ -606,12 +585,80 @@ namespace atomic_dex
             }
         };
 
-        functor_remove(std::move(wallet_cfg_path));
+        if (fs::exists(wallet_cfg_path))
+        {
+            nlohmann::json wallet_config_json_data;
+            std::unordered_set<std::string> active_coins_registry;
+            QFile          fs;
```
Comment (SylEze) [inline src/core/atomicdex/pages/qt.settings.page.cpp:592]: Rename variable to `file`, also might be confusing with `fs` namespace

File: src/core/atomicdex/pages/qt.settings.page.cpp [hunk 5]
```diff
@@ -606,12 +585,80 @@ namespace atomic_dex
...
+            fs.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate);
+            fs.write(QString::fromStdString(wallet_config_json_data.dump(4)).toUtf8());
+            fs.close();
+        }
+
+
+        if (fs::exists(wallet_custom_cfg_path))
+        {
+            nlohmann::json custom_config_json_data;
+            QFile          fs;
```
Comment (SylEze) [inline src/core/atomicdex/pages/qt.settings.page.cpp:638]: Rename variable to `file`, also might be confusing with `fs` namespace

**Review by SylEze — CHANGES_REQUESTED** [review]

**Review by SylEze — APPROVED** [review]

**Comment by tonymorony** [comment]
on OSX app doesn't restart automatically for me after reset (didn't test on other OS yet)

coins list remain the same after restart, so this part works fine

relevant log:


```
[17:01:24] [info] [app.cpp:801] [2702049]: restarting the application
[17:01:24] [error] [app.cpp:815] [2702049]: Couldn't start a new process
[17:01:24] [debug] [app.cpp:651] [2702049]: will quit app, prevent all threading event
[17:01:24] [info] [mm2.service.cpp:295] [2702049]: destroying mm2 service...
[17:01:24] [info] [mm2.service.cpp:299] [2702049]: mm2 signals successfully disconnected
[17:01:24] [info] [mm2.service.cpp:307] [2702049]: processing mm2 stop batch request
[17:01:24] [info] [mm2.service.cpp:303] [2702049]: preparing mm2 stop batch request
[17:01:24] [info] [mm2.service.cpp:310] [2702049]: mm2 stop batch answer received
[17:01:24] [info] [mm2.service.cpp:315] [2702049]: mm2 successfully stopped with rpc stop
[17:01:24] [info] [mm2.service.cpp:346] [2702049]: mm2 init thread destroyed
[17:01:24] [info] [mm2.service.cpp:348] [2702049]: mm2 service fully destroyed
[17:01:24] [info] [main.prerequisites.hpp:216] [2702049]: wally successfully cleaned
[17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
[17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
[17:01:24] [info] [qt.orders.widget.cpp:35] [2702049]: qt_orders widget destroyed
[17:01:24] [info] [coingecko.wallet.charts.cpp:44] [2702049]: coingecko_wallet_charts_service destroyed
```

**Comment by SylEze** [comment]
> on OSX app doesn't restart automatically for me after reset (didn't test on other OS yet)
> 
> coins list remain the same after restart, so this part works fine
> 
> relevant log:
> 
> ```
> [17:01:24] [info] [app.cpp:801] [2702049]: restarting the application
> [17:01:24] [error] [app.cpp:815] [2702049]: Couldn't start a new process
> [17:01:24] [debug] [app.cpp:651] [2702049]: will quit app, prevent all threading event
> [17:01:24] [info] [mm2.service.cpp:295] [2702049]: destroying mm2 service...
> [17:01:24] [info] [mm2.service.cpp:299] [2702049]: mm2 signals successfully disconnected
> [17:01:24] [info] [mm2.service.cpp:307] [2702049]: processing mm2 stop batch request
> [17:01:24] [info] [mm2.service.cpp:303] [2702049]: preparing mm2 stop batch request
> [17:01:24] [info] [mm2.service.cpp:310] [2702049]: mm2 stop batch answer received
> [17:01:24] [info] [mm2.service.cpp:315] [2702049]: mm2 successfully stopped with rpc stop
> [17:01:24] [info] [mm2.service.cpp:346] [2702049]: mm2 init thread destroyed
> [17:01:24] [info] [mm2.service.cpp:348] [2702049]: mm2 service fully destroyed
> [17:01:24] [info] [main.prerequisites.hpp:216] [2702049]: wally successfully cleaned
> [17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
> [17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
> [17:01:24] [info] [qt.orders.widget.cpp:35] [2702049]: qt_orders widget destroyed
> [17:01:24] [info] [coingecko.wallet.charts.cpp:44] [2702049]: coingecko_wallet_charts_service destroyed
> ```

Are you sure it does not happen also on dev ? Because this PR does not modify the app starting logic

**Comment by smk762** [comment]
Seems to be working ok here on mac mini. maybe we need more delay time between shut down and restart on faster systems?

**Comment by tonymorony** [comment]
`Are you sure it does not happen also on dev ? Because this PR does not modify the app starting logic` 

just rechecked - restarting automagically for me on latest dev

`Seems to be working ok here on mac mini. maybe we need more delay time between shut down and restart on faster systems?`

I'm on M1 Max, maybe it makes the difference

**Review by SirSevenG — DISMISSED** [review]
>I'm on M1 Max, maybe it makes the difference

Works just fine on M1 here

**Comment by cipig** [comment]
app does not restart automatically under Linux, but it stops normally and resets the config just fine
it was like that for me since a long time, so not introduced by this PR
i am running the Linux ZIP (unpack, and start `AntaraAtomicDexAppDir/usr/bin/atomicdex-desktop`)
this is the log
```
[21:46:30] [info] [app.cpp:656] [2091940]: Application active
[21:46:38] [warning] [main.prerequisites.hpp:92] [2091940]: qrc:/Dex/App.qml:108:9: QML PopupManager: Binding loop detected for property "height"
[21:46:38] [warning] [main.prerequisites.hpp:92] [2091940]: qrc:/Dex/Settings/SettingModal.qml:308: Error: Invalid write to global property "dialog"
[21:46:44] [info] [app.cpp:653] [2091940]: Application inactive
[21:46:46] [debug] [main.prerequisites.hpp:86] [2091940]: Restarting the application...
[21:46:46] [info] [qt.wallet.manager.cpp:145] [2091940]: Retrieve wallet name: cipinew
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/config/0.5.5.1-coins.cipinew.json
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/0.5.5.1/configs/coins.json
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/0.5.5.1/configs/cfg.ini
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/0.5.5.1/configs/cfg.json
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/logo
[21:46:46] [info] [qt.settings.page.cpp:575] [2091940]: Successfully removed /home/florian/.atomic_qt/themes
[21:46:46] [info] [app.cpp:784] [2091940]: restarting the application
[21:46:46] [error] [app.cpp:798] [2091940]: Couldn't start a new process
[21:46:46] [debug] [app.cpp:635] [2091940]: will quit app, prevent all threading event
[21:46:46] [info] [mm2.service.cpp:295] [2091940]: destroying mm2 service...
[21:46:46] [info] [mm2.service.cpp:299] [2091940]: mm2 signals successfully disconnected
[21:46:46] [info] [mm2.service.cpp:303] [2091940]: preparing mm2 stop batch request
[21:46:46] [info] [mm2.service.cpp:307] [2091940]: processing mm2 stop batch request
[21:46:46] [info] [mm2.service.cpp:310] [2091940]: mm2 stop batch answer received
[21:46:46] [info] [mm2.service.cpp:315] [2091940]: mm2 successfully stopped with rpc stop
[21:46:46] [info] [mm2.service.cpp:346] [2091940]: mm2 init thread destroyed
[21:46:46] [info] [mm2.service.cpp:348] [2091940]: mm2 service fully destroyed
[21:46:46] [info] [main.prerequisites.hpp:216] [2091940]: wally successfully cleaned
[21:46:46] [debug] [main.prerequisites.hpp:86] [2091940]: FatalErrorModal Destructed
[21:46:46] [debug] [main.prerequisites.hpp:86] [2091940]: FatalErrorModal Destructed
[21:46:46] [info] [coingecko.wallet.charts.cpp:44] [2091940]: coingecko_wallet_charts_service destroyed
[21:46:46] [info] [qt.orders.widget.cpp:35] [2091940]: qt_orders widget destroyed
```

the 2 warnings in main.prerequisites.hpp:92 show up when you click on "Reset" and the confirmation window pops up

then the error `[21:46:46] [error] [app.cpp:798] [2091940]: Couldn't start a new process`, maybe it's simply too early, maybe wait a second or two

found a drawback with this approach: if you have coins enabled that don't work because of offline electrums, you can't get rid of those any more, after a reset app will still try to enable them and because they can't be enabled, they don't appear on Portfolio and can't be disabled... will have to try what happens when a coin that is enabled is completely removed/delisted

**Comment by cipig** [comment]
if a coin is delisted and i do a reset, the app crashes on restart
```
[23:27:53] [info] [mm2.service.cpp:215] [2104271]: Retrieving Wallet information of /home/florian/.atomic_qt/config/0.5.5.1-coins.cipinew.json
[23:27:53] [warning] [main.prerequisites.hpp:92] [2104271]: Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt.
You must not let any exception whatsoever propagate through Qt code.
If that is not possible, in Qt 5 you must at least reimplement
QCoreApplication::notify() and catch all exceptions there.

[23:27:53] [debug] [main.prerequisites.hpp:86] [2104271]: FatalErrorModal Destructed
[23:27:53] [debug] [main.prerequisites.hpp:86] [2104271]: FatalErrorModal Destructed
[23:27:53] [error] [main.prerequisites.hpp:98] [2104271]: Object 0x43336f0 destroyed while one of its QML signal handlers is in progress.
Most likely the object was deleted synchronously (use QObject::deleteLater() instead), or the application is running a nested event loop.
This behavior is NOT supported!
qrc:/Dex/Screens/Startup/Login.qml:73: function() { [native code] }
[23:27:53] [error] [main.prerequisites.hpp:106] [2104271]: sigabort received, cleaning mm2
mm2: no process found
[23:27:54] [error] [main.prerequisites.hpp:112] [2104271]: stacktrace:  0# signal_handler(int) at main.cpp:?
 1# 0x00007FF4DCD9FD60 in /lib/x86_64-linux-gnu/libc.so.6
 2# gsignal in /lib/x86_64-linux-gnu/libc.so.6
 3# abort in /lib/x86_64-linux-gnu/libc.so.6
 4# qt_message_handler(QtMsgType, QMessageLogContext const&, QString const&) at main.cpp:?
 5# 0x00007FF4DE228FE0 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
 6# QMessageLogger::fatal(char const*, ...) const in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
 7# QQmlData::destroyed(QObject*) in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Qml.so.5
 8# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
 9# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
10# 0x00007FF4E02F8B66 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
11# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
12# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
13# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
14# 0x00007FF446EC2A5B in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Layouts/libqquicklayoutsplugin.so
15# 0x00007FF446ECD304 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Layouts/libqquicklayoutsplugin.so
16# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
17# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
18# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
19# 0x00007FF4E02FB2B6 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
20# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
21# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
22# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
23# 0x00007FF444395CC6 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Templates.2/libqtquicktemplates2plugin.so
24# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
25# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
26# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
27# 0x00007FF446EC2A5B in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Layouts/libqquicklayoutsplugin.so
28# 0x00007FF446ECD304 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Layouts/libqquicklayoutsplugin.so
29# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
30# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
31# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
32# 0x00007FF4E02F8B66 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
33# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
34# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
35# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
36# 0x00007FF4E02FB2B6 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
37# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
38# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
39# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
40# 0x00007FF4E02F8B66 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
41# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
42# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
43# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
44# 0x00007FF4E02FB2B6 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
45# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
46# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
47# QQuickItem::~QQuickItem() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
48# 0x00007FF4E02F8ADE in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
49# QObjectPrivate::deleteChildren() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
50# QObject::~QObject() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Core.so.5
51# QQuickWindow::~QQuickWindow() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Quick.so.5
52# 0x00007FF444394C61 in /home/florian/Tools/AntaraAtomicDexAppDir/usr/qml/QtQuick/Templates.2/libqtquicktemplates2plugin.so
53# QQmlApplicationEnginePrivate::cleanUp() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Qml.so.5
54# QQmlApplicationEngine::~QQmlApplicationEngine() in /home/florian/Tools/AntaraAtomicDexAppDir/usr/bin/../lib/libQt5Qml.so.5
55# run_app(int, char**) in AntaraAtomicDexAppDir/usr/bin/atomicdex-desktop
56# main in AntaraAtomicDexAppDir/usr/bin/atomicdex-desktop
57# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
58# _start in AntaraAtomicDexAppDir/usr/bin/atomicdex-desktop
```

**Comment by smk762** [comment]
> found a drawback with this approach: if you have coins enabled that don't work because of offline electrums, you can't get rid of those any more, after a reset app will still try to enable them and because they can't be enabled, they don't appear on Portfolio and can't be disabled... will have to try what happens when a coin that is enabled is completely removed/delisted

In this event, perhaps we should just mark it as disabled

> if a coin is delisted and i do a reset, the app crashes on restart

Good catch, will have a look

**Review by cipig — DISMISSED** [review]
works fine now, coin is removed completely in case of delisting

**Review by cipig — DISMISSED** [review]

**Review by SirSevenG — CHANGES_REQUESTED** [review]
Tested on M1 only so far, will recheck on Win11 soon. Still, faced 2 issues here:

1. Updating to latest commit from other build will reset users' coin on 1st start

- on a fresh wallet on other app build enable N coins, close app
- update app to latest commit of this branch
- start app
- Only 3 default coins are activated

2. After enabling N coins and restarting app, M < N coins will be activated on startup

- Activate N(ae: 9) coins (12 should be activated in total)
- Restart app, simply by closing it
- After startup check dashboard - coins total: not all coins previously activated are active after app restart (10 or 11 will be active)

Don't reproduce on latest release, no electrum or other errors in mm2 log, app log simply ignores should-be-active coins. Didn't see anything relevant.

**Comment by smk762** [comment]
Thanks for the feedback @SirSevenG 
I've found where the incomplete coin enabling issue was occurring, though might need @SylEze to assist with resolution.
Flow is:
- coins file is read into json
- tickers to enable update json "active"
- json writes to file.
problem is that it seems this happens in parallel so it reads too early before last updates done and only last enabling tickers are actually updated in file.
I tested using [QSaveFile](https://doc.qt.io/qt-5/qsavefile.html#commit) which half fixed it (no more logs about parsing errors when reading a partially written file), but the early read issue persists.

I assume a lock to avoid reading file until clear is needed. Alternatively, it might be better to use the coins_registry to write out.
Will set this PR back to draft, and revisit for 0.5.7

**Comment by tonymorony** [comment]
unattached reviewers for now
@smk762 please re-request review when this PR is ready for another round of testing

**Review by cipig — APPROVED** [review]
still works fine

**Review by Canialon — APPROVED** [review]
What a lovely PR 💘

**Review by SirSevenG — APPROVED** [review]
Works for me

**Comment by smk762** [comment]
> on OSX app doesn't restart automatically for me after reset (didn't test on other OS yet)
> 
> coins list remain the same after restart, so this part works fine
> 
> relevant log:
> 
> ```
> [17:01:24] [info] [app.cpp:801] [2702049]: restarting the application
> [17:01:24] [error] [app.cpp:815] [2702049]: Couldn't start a new process
> [17:01:24] [debug] [app.cpp:651] [2702049]: will quit app, prevent all threading event
> [17:01:24] [info] [mm2.service.cpp:295] [2702049]: destroying mm2 service...
> [17:01:24] [info] [mm2.service.cpp:299] [2702049]: mm2 signals successfully disconnected
> [17:01:24] [info] [mm2.service.cpp:307] [2702049]: processing mm2 stop batch request
> [17:01:24] [info] [mm2.service.cpp:303] [2702049]: preparing mm2 stop batch request
> [17:01:24] [info] [mm2.service.cpp:310] [2702049]: mm2 stop batch answer received
> [17:01:24] [info] [mm2.service.cpp:315] [2702049]: mm2 successfully stopped with rpc stop
> [17:01:24] [info] [mm2.service.cpp:346] [2702049]: mm2 init thread destroyed
> [17:01:24] [info] [mm2.service.cpp:348] [2702049]: mm2 service fully destroyed
> [17:01:24] [info] [main.prerequisites.hpp:216] [2702049]: wally successfully cleaned
> [17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
> [17:01:24] [debug] [main.prerequisites.hpp:86] [2702049]: FatalErrorModal Destructed
> [17:01:24] [info] [qt.orders.widget.cpp:35] [2702049]: qt_orders widget destroyed
> [17:01:24] [info] [coingecko.wallet.charts.cpp:44] [2702049]: coingecko_wallet_charts_service destroyed
> ```

Can you please check if this still happens for you @tonymorony ?

**Review by SirSevenG — DISMISSED** [review]
Rechecked on Win 11 and Mac, app restarts correctly, coins are up including zhtlc.

**Review by cipig — APPROVED** [review]

**Review by Canialon — APPROVED** [review]
Works good on Windows 10 and Mac
noticed that ARRR/ZOMBIE will start activation again if they were in activation process, nice:3

**Comment by smk762** [comment]
> ^^

Can you please recheck this one @tonymorony ?

**Review by tonymorony — APPROVED** [review]

**Pull request merged**